src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs

```diff
@@ -1386,5 +1388,136 @@ private IIsPatternExpression CreateBoundIsPatternExpressionOperation(BoundIsPatt
+        {
+            switch(boundQueryClause.Syntax.Kind())
+            {
+                case SyntaxKind.QueryExpression:
+                    return CreateBoundQueryExpressionOperation(boundQueryClause);
+
+                case SyntaxKind.QueryContinuation:
+                    return CreateBoundQueryContinuationOperation(boundQueryClause);
+
+                case SyntaxKind.AscendingOrdering:
+                    return CreateBoundOrderingExpressionOperation(boundQueryClause, OrderKind.Ascending);
+
+                case SyntaxKind.DescendingOrdering:
+                    return CreateBoundOrderingExpressionOperation(boundQueryClause, OrderKind.Descending);
+
+                case SyntaxKind.QueryBody:
+                    return boundQueryClause.Value != null ? Create((BoundQueryClause)boundQueryClause.Value) : null;
+
+                default:
+                    throw ExceptionUtilities.Unreachable;
+            }
+        }
+
+        private IQueryExpression CreateBoundQueryExpressionOperation(BoundQueryClause boundQueryClause)
+        {
+            Lazy<IOperation> lastClauseOrContinuation = new Lazy<IOperation>(() => Create(boundQueryClause.Value));
+            SyntaxNode syntax = boundQueryClause.Syntax;
+            ITypeSymbol type = boundQueryClause.Type;
+            Optional<object> constantValue = ConvertToOptional(boundQueryClause.ConstantValue);
+            return new LazyQueryExpression(lastClauseOrContinuation, _semanticModel, syntax, type, constantValue);
+        }
+
+        private IQueryContinuation CreateBoundQueryContinuationOperation(BoundQueryClause boundQueryClause)
+        {
+            Lazy<IOperation> queryBody = new Lazy<IOperation>(() => Create(boundQueryClause.Value));
+            IRangeVariableSymbol definedSymbol = boundQueryClause.DefinedSymbol;
+            SyntaxNode syntax = boundQueryClause.Syntax;
+            ITypeSymbol type = boundQueryClause.Type;
+            Optional<object> constantValue = ConvertToOptional(boundQueryClause.ConstantValue);
+            return new LazyQueryContinuation(queryBody, definedSymbol, _semanticModel, syntax, type, constantValue);
+        }
+
+        private IOrderingExpression CreateBoundOrderingExpressionOperation(BoundQueryClause boundQueryClause, OrderKind orderKind)
+        {
+            Lazy<IOperation> expression = new Lazy<IOperation>(() => Create(boundQueryClause.Value));
+            SyntaxNode syntax = boundQueryClause.Syntax;
+            ITypeSymbol type = boundQueryClause.Type;
+            Optional<object> constantValue = ConvertToOptional(boundQueryClause.ConstantValue);
+            return new LazyOrderingExpression(expression, orderKind, _semanticModel, syntax, type, constantValue);
+        }
+
+        private QueryClauseKind? GetQueryClauseKind(BoundQueryClause boundQueryClause)
+        {
+            switch(boundQueryClause.Syntax.Kind())
```

Comment:
Sure, I will pull all of the syntax kind comparisons into a single switch statement.